pkg/controller/route/route_controller.go

```diff
@@ -137,34 +138,47 @@ func (rc *RouteController) reconcileNodeRoutes() error {
 }
 
 func (rc *RouteController) reconcile(nodes []*v1.Node, routes []*cloudprovider.Route) error {
-	// nodeCIDRs maps nodeName->nodeCIDR
-	nodeCIDRs := make(map[types.NodeName]string)
+	// for each node a map of podCIDRs and their created status
+	var l sync.Mutex
+	nodeRoutesStatuses := make(map[types.NodeName]map[string]bool)
 	// routeMap maps routeTargetNode->route
-	routeMap := make(map[types.NodeName]*cloudprovider.Route)
+	routeMap := make(map[types.NodeName][]*cloudprovider.Route)
 	for _, route := range routes {
 		if route.TargetNode != "" {
-			routeMap[route.TargetNode] = route
+			routeMap[route.TargetNode] = append(routeMap[route.TargetNode], route)
 		}
 	}
 
 	wg := sync.WaitGroup{}
 	rateLimiter := make(chan struct{}, maxConcurrentRouteCreations)
+	// searches existing routes by node for a matching route
 
 	for _, node := range nodes {
 		// Skip if the node hasn't been assigned a CIDR yet.
-		if node.Spec.PodCIDR == "" {
+		if len(node.Spec.PodCIDRs) == 0 {
 			continue
 		}
 		nodeName := types.NodeName(node.Name)
-		// Check if we have a route for this node w/ the correct CIDR.
-		r := routeMap[nodeName]
-		if r == nil || r.DestinationCIDR != node.Spec.PodCIDR {
-			// If not, create the route.
+		l.Lock()
+		nodeRoutesStatuses[nodeName] = make(map[string]bool)
+		l.Unlock()
+		// for every node, for every cidr
+		for _, podCIDR := range node.Spec.PodCIDRs {
+			// we add it to our nodeCIDRs map here because add and delete go routines run at the same time
+			nodeRoutesStatuses[nodeName][podCIDR] = false
```

Comment:
added for clarity and future maintenance, though locks at like 168, 174 are actually not needed because at this point nothing will be editing that key/slice pair.

Reply:
Ah, right, as far as the `nodeRoutesStatuses` map is concerned, the only locking needed is that needed to keep apart the write `nodeRoutesStatuses[nodeName] = make(map[string]bool)` and the reads in the goroutines that invoke `rc.routes.CreateRoute`.

However, there is also a mutual exclusion problem for the inner maps, the one for each nodeName that maps podCIDR to bool.  These inner maps get writes in the aforementioned goroutines and also in the mainline code that launches them.  For one particular nodeName it is possible that a goroutine tries to write into the inner map concurrently with a write by the mainline code or a read in `nodeHasCidr`.  Currently all those but one lock the mutex; they should all do so.